components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodProjectManager.kt

```diff
@@ -40,8 +42,16 @@ class GitpodProjectManager(
         application.invokeLaterOnWriteThread {
             application.runWriteAction {
                 try {
-                    ProjectJdkTable.getInstance().preconfigure()
-                    pendingSdk.complete(ProjectJdkTable.getInstance().allJdks.firstOrNull())
```

Comment:
Should not it allredy return all jdks we need to loop and find one whcih is from java home?

There is also `JavaHomeFinder` can it be helpful? i.e. instead of trying to parse JAVA_HOME themself?  It seems to have different strategies: https://github.com/JetBrains/intellij-community/blob/379444cdda4dc4a2981df41c94ab9998ab050850/platform/lang-impl/src/com/intellij/openapi/projectRoots/impl/JavaHomeFinderBasic.java#L41

Reply:
> Should not it allredy return all jdks we need to loop and find one whcih is from java home?

Unfortunately, only Java 17 is in the list at the first start of the workspace (the other Detected SDKs only appear there after the project is already loaded) (so only at the restart of the workspace would the list contain both Java 17 and 11). I'm looking for a way to make this detection of sdks available to happen before we loop through `jdkTable.allJdks`.

> There is also JavaHomeFinder can it be helpful?

I tried it too, but it only useful info it returns is the homePath (as string) of each detect SDK.
There's no info about which one is marked as default as per `JAVA_HOME`.
That's why I'm still querying for the environment variable.

Reply:
> I tried it too, but it only useful info it returns is the homePath (as string) of each detect SDK. There's no info about which one is marked as default as per JAVA_HOME. That's why I'm still querying for the environment variable.

Looking at the code first one should be if you disable [[1](https://github.com/JetBrains/intellij-community/blob/379444cdda4dc4a2981df41c94ab9998ab050850/platform/lang-impl/src/com/intellij/openapi/projectRoots/impl/JavaHomeFinderBasic.java#L33-L35)] these flags

Reply:
> Looking at the code first one should be if you disable [[1](https://github.com/JetBrains/intellij-community/blob/379444cdda4dc4a2981df41c94ab9998ab050850/platform/lang-impl/src/com/intellij/openapi/projectRoots/impl/JavaHomeFinderBasic.java#L33-L35)] these flags

Tried it too. Disabled all the flags but it was still returning an array with the 3 available SDK paths from `/home/gitpod/.sdkman/candidates/java/`.

Note also that none of the methods from `JavaHomeFinderBasic` return an SDK object. They all return paths, so I couldn't find a use for this class.

Reply:
Could you post values? It should lookup from PATH and then from JAVA_HOME. I would expect that PATH has default in too.

Reply:
Yes, these are the logs returned by `JavaHomeFinder.getFinder().findExistingJdks().forEach { thisLogger().warn(it) }`
```
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/11.0.16.fx-zulu
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/17.0.4.fx-zulu
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/current
```
Yes, these are the logs returned by `JavaHomeFinder.getFinder().checkDefaultInstallDir(false).checkUsedInstallDirs(false).checkConfiguredJdks(false).findExistingJdks().forEach { thisLogger().warn(it) }`
```
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/11.0.16.fx-zulu
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/17.0.4.fx-zulu
WARN - #i.g.j.r.GitpodProjectManager - /home/gitpod/.sdkman/candidates/java/current
```

But having these paths doesn't help in this case because it doesn't indicate which one is the one from $JAVA_HOME, and it also doesn't return an SDK object; only strings are returned.

I need an SDK object for adding it to the table `ProjectJdkTable.getInstance().addJdk(sdk)`, because I can't call `ProjectRootManager.getInstance(project).projectSdk = sdk` if it's not added to ProjectJdkTable first.

Reply:
I think taking first one here it a right choice because they are computed from PATH env var, i.e. the same way how to it is computed in the shell.

Reply:
/home/gitpod/.sdkman/candidates/java/11.0.16.fx-zulu is in fact correct

Reply:
Gotcha. I'll take from this list instead of querying from the environment var then.